[WIP] Sparse Tensor

eric-haibin-lin opened the pull request:
Please don't merge in. This PR includes some changes to enable imperative execution for sparse ndarrays for a simple elemwise_add operator on cpu. 

nnvm changes at https://github.com/dmlc/nnvm/pull/110

@piiswrong @mli @madjam @reminisce

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -344,27 +450,104 @@ class NDArray {
+    /*! \brief storage handle from storage engine */
     Storage::Handle shandle;
+    /*! \brief storage handles for aux data (e.g index) */
+    std::vector<Storage::Handle> aux_handles;
```
vector is probably overkill

eric-haibin-lin replied:
CSR needs 2 aux data. I can make it an array of size 2

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -344,27 +450,104 @@ class NDArray {
+    std::vector<Storage::Handle> aux_handles;
     /*! \brief variable from engine */
     Engine::VarHandle var;
+    std::vector<Engine::VarHandle> aux_vars;
```
no need for additional var

piiswrong commented on include/mxnet/op_attr_types.h:
```diff
@@ -61,6 +61,19 @@ using FCompute = std::function<void (const nnvm::NodeAttrs& attrs,
+ * TODO should probably change const std::vector<NDArray>& outputs to
+                               std::vector<NDArray> *outputs
+ */
+using FComputeND = std::function<void (const nnvm::NodeAttrs& attrs,
```
we can probably just call it FComputeEx

piiswrong commented on python/mxnet/ndarray.py:
```diff
@@ -560,24 +559,6 @@ def wait_to_read(self):
 
 
-    @property
-    def shape(self):
```
?

eric-haibin-lin replied:
I plan to move all shared properties/method to NDArrayBase.

piiswrong replied:
can SparseNDArray inherit NDArray instead of NDArrayBase?

eric-haibin-lin replied:
That means for any new function added to NDArray, we need to either implement it for SparseNDArray or disable it in SparseNDArray. Otherwise it confuses user with an inherited method which doesn't work

piiswrong replied:
any such examples? NDArray interface has been fairly stable.
The problem with SparseNDArray not being a NDArray is all the places where you use isinstance(x, NDArray) need to be changed to isinstance(x, NDArrayBase)

eric-haibin-lin replied:
If the interface is stable, that's fine. I just want to make sure user doesn't make the wrong function call from SparseND which is directed to NDArray and doesn't work

reminisce commented on include/mxnet/ndarray.h:
```diff
@@ -67,20 +91,66 @@ class NDArray {
       Mkl_mem_ = std::make_shared<MKLMemHolder>();
 #endif
   }
+  NDArray(NDArray data, std::vector<NDArray> aux_data, Context ctx, NDArrayChunkType chunk_type, const TShape &shape)
```
Why is aux_data passed by value?

eric-haibin-lin replied:
It doesn't matter. NDArray has the pointer to the chunk

reminisce replied:
The point is to avoid copying entire vector.

eric-haibin-lin replied:
sure

reminisce commented on include/mxnet/ndarray.h:
```diff
@@ -374,31 +557,65 @@ class NDArray {
+        aux_handles.push_back(Storage::Get()->Alloc(aux_bytes, ctx));
+        delay_alloc = false;
+        // Initialize aux_shape and shape
+        this->aux_shapes = aux_shapes;
```
this->aux_shapes = std::move(aux_shapes)?

eric-haibin-lin replied:
To save memory ops?

reminisce replied:
Same as the above comment, avoid copying the entire vector.

reminisce commented on src/operator/tensor/elemwise_binary_op.h:
```diff
@@ -32,6 +34,105 @@ void BinaryCompute(const nnvm::NodeAttrs& attrs,
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  auto indices_l = nd_l.aux_data(0).FlatTo1D<xpu, ROW_SPARSE_TYPE>(s);
+  auto indices_r = nd_r.aux_data(0).FlatTo1D<xpu, ROW_SPARSE_TYPE>(s);
+  auto indices_out = output.aux_data(0).FlatTo1D<xpu, ROW_SPARSE_TYPE>(s);
```
A general comment is that since this implements row sparse format of tensors, can we have meaningful interfaces to get the number of rows and row index array from a NDArray? Currently, these two attributes are retrieved from `NDArray::aux_shape(0)[0]` and `NDArray::aux_data(0).FlatTo1D<xpu, ROW_SPARSE_TYPE>(s)`, which seems not very natural. I would have to look into the detailed implementation of Chunk::CheckAndAlloc to understand the meaning of these.

eric-haibin-lin replied:
Good point. I'll add some interface in NDArray to abstract this out

piiswrong commented:
please run `git rebase upstream/master` to clean up diff

piiswrong commented on include/mxnet/io.h:
```diff
@@ -44,6 +44,19 @@ class IIterator : public dmlc::DataIter<DType> {
+ * \param DType data type
+ */
+template<typename DType>
+class SparseIIterator : public IIterator<DType> {
```
Put this undr src/ don't expose to user

eric-haibin-lin replied:
ok

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -28,8 +28,22 @@
+
+namespace ndarray {
+template<typename from_xpu, typename to_xpu>
+void Copy(const TBlob &from, TBlob *to, Context from_ctx, Context to_ctx, RunContext ctx);
```
Put these undr src/

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -53,6 +67,27 @@ class AGNodeEntry {
+// enum for storage types
+#define CSR_IND_PTR_TYPE mshadow::kInt32
+#define CSR_IDX_DTYPE mshadow::kInt32
+#define ROW_SPARSE_IDX_TYPE mshadow::kInt32
```
One type is enough?
use typedef instead of define

eric-haibin-lin replied:
yeah I'm thinking about default to 64bit for row_sparse::idx and csr::indices, and 32bit for csr::indptr. should be sufficient

piiswrong replied:
Let's keep them the same

eric-haibin-lin replied:
ok

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -73,41 +108,152 @@ class NDArray {
    *  make sure the memory region is available through out the life of NDArray
    * \param data the memory content of static data
    * \param dev_id the device id this tensor sits at
+   * \param shared_var the same var handle shared with others.
```
why

eric-haibin-lin replied:
mainly used for https://github.com/eric-haibin-lin/mxnet/blob/master/include/mxnet/ndarray.h#L464 to ensure thread safety when the frontend retrieves the read-only `indices` and `indptr` of the sparse ndarray

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -73,41 +108,152 @@ class NDArray {
+   * nnz that is smaller than nnz1+nnz2. Therefore, the storage shape's size
+   * needs to be shrunk from nnz1+nnz2 to nnz.
+   */
+  inline void SetStorageShape(const TShape& sshape) {
```
set_storage_shape

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -73,41 +108,152 @@ class NDArray {
+   * for the final result. After the operation is done, the exact size of
+   * the shape is known and need to be reset using this function.
+   */
+  inline void SetAuxShape(size_t i, const TShape& shape) const {
```
set_aux_shape

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -119,6 +265,28 @@ class NDArray {
+  /*!
+   * \return the number of aux data used for given storage type
+   */
+  static size_t NumAuxData(NDArrayStorageType stype) {
```
Don't like this being a member function very much

eric-haibin-lin replied:
maybe move it to a util file?

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -260,17 +440,38 @@ class NDArray {
+   * \param end end index in first dim (exclusive)
+   * \return sliced NDArray
+   */
+  void SliceEx(index_t begin, index_t end, NDArray *dst) const;
```
why?

eric-haibin-lin replied:
I can stick to the existing `Slice()` interface instead of creating a new one

piiswrong commented on include/mxnet/ndarray.h:
```diff
@@ -312,8 +514,25 @@ class NDArray {
+    CHECK_NE(storage_type(), kDefaultStorage);
+    ptr_->CheckAndAllocData(storage_shape, dtype_);
+  }
+  inline void CheckAndAllocAuxData(size_t i, const TShape &aux_shape) const {
```
why separate these?

eric-haibin-lin replied:
In some case the sparse operator doesn't know how much memory to allocate at once. e.g. https://github.com/eric-haibin-lin/mxnet/blob/master/src/operator/tensor/elemwise_unary_op.h#L308 
we have to allocate some temp memory to work out the aux_shape of the result. This makes it easier to reuse allocated temp memory explicitly, but makes code harder to read.
Another way is to just request for a temp memory to just calculate the aux_shape, and leave efficient memory management to the storage manager instead.

piiswrong commented on include/mxnet/op_attr_types.h:
```diff
@@ -18,6 +17,9 @@
 #include "./operator.h"
 #include "./ndarray.h"
 
+#define FCOMP_EX_CPU "FComputeEx<cpu>"
```
Write FComputeEx<cpu> directly

eric-haibin-lin replied:
ok

piiswrong commented on include/mxnet/op_attr_types.h:
```diff
@@ -61,6 +63,17 @@ using FCompute = std::function<void (const nnvm::NodeAttrs& attrs,
+ *       Dispatched only when operators process non-default storage inputs or outputs
+ */
+using FComputeEx = std::function<void (const nnvm::NodeAttrs& attrs,
+                                     const OpContext& ctx,
```
may need a different context representation

eric-haibin-lin replied:
why?

piiswrong commented on python/mxnet/kvstore.py:
```diff
@@ -48,7 +48,7 @@ def updater_handle(key, lhs_handle, rhs_handle, _):
 class KVStore(object):
     """A key-value store for synchronization of values, over multiple devices."""
-    def __init__(self, handle):
+    def __init__(self, handle, name2idx=None):
```
let's discuss this

eric-haibin-lin replied:
Yeah I am not sure if this is the best place to hold it. This is actually not required for the sparse feature. I can remove it temporarily for now.

eric-haibin-lin replied:
https://github.com/dmlc/mxnet/pull/6765

piiswrong commented on python/mxnet/module/module.py:
```diff
@@ -398,7 +399,7 @@ def bind(self, data_shapes, label_shapes=None, for_training=True,
             assert self._arg_params is None and self._aux_params is None
             param_arrays = [
-                nd.zeros(x[0].shape, dtype=x[0].dtype)
+                sparse_nd.zeros(x[0].storage_type, x[0].shape, dtype=x[0].dtype)
```
?

piiswrong commented on python/mxnet/ndarray.py:
```diff
@@ -2406,37 +2437,5 @@ def %s(%s):
 
-
-# pylint: enable=too-many-locals, invalid-name
-def _init_ndarray_module(ndarray_class, root_namespace):
```
?

piiswrong commented on python/mxnet/optimizer.py:
```diff
@@ -332,7 +333,8 @@ def create_state(self, index, weight):
             return None
         else:
-            return zeros(weight.shape, weight.context, dtype=weight.dtype)
+            return sparse_zeros(weight.storage_type, weight.shape,
```
its better to add a storage_type argument to zeros and make it output sparse arrays accordingly.

eric-haibin-lin replied:
Good point. Will change this.

reminisce replied:
We discussed this last Friday. In order to do this, we need to move all the util functions supporting sparse ndarrays in ndarray.py into an ndarray util python file to avoid circular import between ndarray.py and sparse_ndarray.py. Quite a few files calling those functions will be affected.

piiswrong commented on src/c_api/c_api.cc:
```diff
@@ -287,6 +320,16 @@ int MXNDArraySlice(NDArrayHandle handle,
   API_END_HANDLE_ERROR(delete ptr);
 }
 
+int MXNDArraySliceEx(NDArrayHandle handle,
```
Why not slice and assign?

eric-haibin-lin replied:
does slice and assign still share the memory?

piiswrong commented on src/c_api/c_api_ndarray.cc:
```diff
@@ -242,23 +275,61 @@ void PushFCompute(const FCompute& fn,
+        CastNonDefaultStorage<gpu>(ndoutputs, temp_out, opctx);
+        rctx.get_stream<gpu>()->Wait();
+#else
+        LOG(FATAL) << MXNET_GPU_NOT_ENABLED_ERROR;
```
this can be done before push

eric-haibin-lin replied:
you mean the call to `GetDefaultBlobs` before push? It's potentially cast the storage type, which is an operator

piiswrong commented on src/executor/attach_op_execs_pass.cc:
```diff
@@ -135,52 +161,78 @@ class BackwardOpExecutor : public OpExecutor {
+    fcompute_(attrs_, op_ctx, in_data_, req, out_data_);
+  }
+  void Setup() override {
+    in_data_ = in_array;
```
where is in_array defined?

eric-haibin-lin replied:
in https://github.com/dmlc/mxnet/blob/master/src/executor/exec_pass.h#L30

piiswrong commented on src/executor/attach_op_execs_pass.cc:
```diff
@@ -24,9 +28,33 @@ namespace exec {
+      op_->Forward(op_ctx, in_data_, req, out_data_, aux_data_);
+      CastNonDefaultStorage<gpu>(out_array, temp_out_, op_ctx);
+#else
+      LOG(FATAL) << MXNET_GPU_NOT_ENABLED_ERROR;
```
Use LOG(DEBUG) and change it to only print in debug build

eric-haibin-lin replied:
I was not aware LOG(DEBUG) was used anywhere in the code..

eric-haibin-lin commented on include/mxnet/c_api.h:
```diff
@@ -246,6 +246,38 @@ MXNET_DLL int MXNDArrayCreateEx(const mx_uint *shape,
+ * \return 0 when success, -1 when failure happens
+ */
+MXNET_DLL int MXNDArrayCreateSparseEx(int storage_type,
+                    const mx_uint *shape,
```
identation

eric-haibin-lin commented on include/mxnet/c_api.h:
```diff
@@ -1003,6 +1063,25 @@ MXNET_DLL int MXSymbolInferType(SymbolHandle sym,
+/*!
+ * \brief infer storage type of unknown input types given the known one.
+ */
+MXNET_DLL int MXSymbolInferStorageType(SymbolHandle sym,
```
This API could probably be removed

eric-haibin-lin commented on include/mxnet/ndarray.h:
```diff
@@ -29,7 +29,11 @@
+namespace ndarray {
+template<typename from_xpu, typename to_xpu>
+void Copy(const TBlob &from, TBlob *to, Context from_ctx, Context to_ctx, RunContext ctx);
+};
```
No need for this forward declaration

eric-haibin-lin commented on include/mxnet/op_attr_types.h:
```diff
@@ -61,6 +60,17 @@ using FCompute = std::function<void (const nnvm::NodeAttrs& attrs,
+                                     const OpContext& ctx,
+                                     const std::vector<NDArray>& inputs,
+                                     const std::vector<OpReqType>& req,
+                                     const std::vector<NDArray>& outputs)>;
```
indentation

eric-haibin-lin commented on python/mxnet/ndarray.py:
```diff
@@ -949,6 +975,13 @@ def backward(self, out_grad=None, retain_graph=False):
             c_array(NDArrayHandle, ograd_handles),
             ctypes.c_int(retain_graph)))
 
+    def to_csr(self):
```
documentation

eric-haibin-lin commented on python/mxnet/sparse_ndarray.py:
```diff
@@ -0,0 +1,576 @@
+    def _sync_copyfrom(self, source_array):
+        raise Exception('Not implemented for SparseND yet!')
+
+    def _slice(self, start, stop):
```
remove shared mem slice for sparse ndarray. always use crop and assign

eric-haibin-lin commented on python/mxnet/sparse_ndarray.py:
```diff
@@ -0,0 +1,576 @@
+        raise Exception('Transpose is not supported for SparseNDArray.')
+
+    @property
+    def aux_types(self):
```
_aux_types

eric-haibin-lin commented on python/mxnet/sparse_ndarray.py:
```diff
@@ -0,0 +1,576 @@
+    row i are stored in indices[indptr[i]:indptr[i+1]] and their corresponding values are stored
+    in values[indptr[i]:indptr[i+1]].
+
+    """
```
add example

eric-haibin-lin commented on python/mxnet/sparse_ndarray.py:
```diff
@@ -0,0 +1,576 @@
+
+    RowSparseNDArray is used principally in the definition of gradients for operations
+    that have sparse gradients (e.g. SparseEmbedding).
+    """
```
add example

eric-haibin-lin commented on python/mxnet/symbol.py:
```diff
@@ -721,6 +723,89 @@ def list_auxiliary_states(self):
             self.handle, ctypes.byref(size), ctypes.byref(sarr)))
         return [py_str(sarr[i]) for i in range(size.value)]
 
+    def infer_storage_type(self, *args, **kwargs):
```
should include ctx. or could just be removed

eric-haibin-lin commented on python/mxnet/test_utils.py:
```diff
@@ -657,14 +715,29 @@ def check_symbolic_backward(sym, location, out_grads, expected, rtol=1e-5, atol=
     args_grad_npy = {k:_rng.normal(size=v.shape) for k, v in expected.items()}
-    args_grad_data = {k: mx.nd.array(v, ctx=ctx) for k, v in args_grad_npy.items()}
+    # args_grad_data should be casted to storage type if hinted
+    # TODO(haibin) this is a temporary solution for testing. remove later
```
verify if it's still needed

eric-haibin-lin commented on src/c_api/c_api.cc:
```diff
@@ -154,6 +154,39 @@ int MXNDArrayCreateEx(const mx_uint *shape,
 }
 
+int MXNDArrayCreateSparseEx(int storage_type,
+                    const mx_uint *shape,
```
indentation

eric-haibin-lin commented on src/executor/attach_op_execs_pass.cc:
```diff
@@ -62,12 +86,14 @@ class ForwardOpExecutor : public OpExecutor {
-  void Run(RunContext rctx) override {
+  void Run(RunContext rctx, bool is_gpu) override {
+    // TODO(haibin) support storage fallback for BackwardOpExecutor
     op_ctx.run_ctx = rctx;
```
support cast storage for BackwardOpExecutor

eric-haibin-lin commented on src/kvstore/comm.h:
```diff
@@ -64,11 +68,18 @@ class CommCPU : public Comm {
     nthread_reduction_ = dmlc::GetEnv("MXNET_KVSTORE_REDUCTION_NTHREADS", 4);
     bigarray_bound_ = dmlc::GetEnv("MXNET_KVSTORE_BIGARRAY_BOUND", 1000 * 1000);
+    // TODO(junwu) delete the following data member, now for benchmark only
+    is_serial_push_ = dmlc::GetEnv("MXNET_KVSTORE_SERIAL_PUSH", 0);
```
remove after the benchmark

eric-haibin-lin commented on src/ndarray/ndarray.cc:
```diff
@@ -56,40 +61,101 @@ NDArray NDArray::Reshape(const TShape &shape) const {
-      op, attrs, &inputs, &outputs);
-    return outputs[0];
-  } else {
+  CHECK_NE(storage_type(), kUndefinedStorage);
```
remove shared memory implementation

eric-haibin-lin commented:
### A list of frontend changes in this PR

Note: rsp is short for row_sparse
- Frontend changes: 
  NDArrayBase
  -> NDArray
  -> SparseNDArray
  -> CSRNDArray / RowSparseNDArray 

- All `NDArray(NDArrayHandle(handle))` is replaced by `_ndarray_cls(NDArrayHandle(handle))` to construct NDArray/CSRNDArray/RowSparseNDArray based on storage type
- Added `to_rsp`, `to_csr` and `to_dense` method for NDArray storage type conversion
- Implemented `zeros`, `load`, `save` for SparseNDArray, functions are moved to ndarray_utils 
- `SparseNDArray` only supports [:] for assignment
- support `__getitem__` with rows for CSRNDArray, but not for RowSparseNDArray
- `__iadd__`, `__imul__`, `__idiv___`, `__isub__` is not supported for SparseNDArray. `__add__`, `__mul__`, etc are inherited from NDArray
- `at`, `reshape`, `broadcast_to`, `T` not supported for SparseNDArray yet
- user is able to retrieve aux (dense) NDArray via `indices`, `values` and `indptr` properties (`indptr` is only available for CSRNArray)
- `Variable` has one more attr called `__storage_type__`
- List of operators implemented (CPU only - many of them shares templated implementation but don't have great performance on GPU):
  `elemwise_add` (row_sparse)
  `cast_storage` (csr<->dense, rsp<->dense, forward only) 
  `dot` (csr dot dense, csr.T dot dense, csr dot row_sparse, csr.T dot row_sparse) 
  `sparse_embedding` (does the same thing as dot, expect weight is auto-allocated)
  `sparse_retain` (see https://github.com/dmlc/mxnet/pull/5800/files#diff-1b545c728261fd4bf9d82d1340ea7a63R275) 
  `sgd_update` & `sgd_mom_update` (only updating weights with non-zeros rows in gradient, when either weight or gradient is row_sparse)
- `LibSVMIter` reads libsvm format input and output `CSRNDArray`  
- `NDArrayIter` can accept `CSRNDArray` as input 
- kvstore is able to initialize `row_sparse` weight

TODO list
- kvstore API needs to adopt changes in #6765
- remove infer_storage_type python interface 
-  Add test for multi-dimensional RowSparse NDArray
- remove shared memory `_slice` for SparseNDArray, always perform slice and assign
- merge unary & binary operators from Chris's branch
- bug fix for `NDArrayIter` and `LibSVMIter` (https://github.com/eric-haibin-lin/mxnet/issues/110 https://github.com/eric-haibin-lin/mxnet/issues/98)
- Plus the ones I mentioned in the previous review

eric-haibin-lin commented:
TODO list for next PR:
- timeit the cost for c_api invocation for NDArray constructor (a + b & c_api)
- _to_rsp, _to_csr. see if scipy removes zeros
- move ndarray_utils, nd, sparse_nd to a folder 
- throw exception for not implemented methods for in sparse_nd
- rename: value -> data 
- asnumpy() returns scipy object
- to_dense -> todense
- rename __storage_type__
- remove sparse_embedding
- remove aux type switch -> MSHADOW_IDX_SWITCH
